### PR TITLE
Add options for controling set request encoding

### PR DESCRIPTION
**Pull request by DanG100**

### CODE REVIEW

File: ygnmi/ygnmi.go [hunk 1]
```diff
@@ -210,6 +212,22 @@ func WithEncoding(enc gpb.Encoding) Option {
...
 
+// WithSetPreferProtoEncoding creates an option that prefers encoding SetRequest using proto encoding.
+// This option is only relevant for Update and Replace.
+func WithSetPreferProtoEncoding() Option {
+	return func(o *opt) {
+		o.preferProto = true
+	}
+}
+
+// WithSetFallbackEncoding creates an option that fallback to encoding SetRequests with JSON or an Any proto.
```
Comment (wenovus) [inline ygnmi/ygnmi.go:223]: explain under what circumstances is the fallback encoding used?

File: ygnmi/gnmi.go [hunk 2]
```diff
@@ -330,25 +333,49 @@ const (
...
-		// Since the GoStructs are generated using preferOperationalState, we
-		// need to turn on preferShadowPath to prefer marshalling config paths.
-		js, err := ygot.Marshal7951(val, ygot.JSONIndent("  "), &ygot.RFC7951JSONConfig{AppendModuleName: true, PreferShadowPath: preferShadowPath})
-		if err != nil {
-			return fmt.Errorf("could not encode value into JSON format: %w", err)
-		}
 		update := &gpb.Update{
 			Path: path,
-			Val:  &gpb.TypedValue{Value: &gpb.TypedValue_JsonIetfVal{JsonIetfVal: js}},
 		}
```
Comment (wenovus) [inline ygnmi/gnmi.go:348]: It seems to me it's cleaner to put this at the end. There seems to be multiple overwritings of update.Val, which makes it a bit more difficult to read.

File: ygnmi/gnmi.go [hunk 3]
```diff
@@ -330,25 +333,49 @@ const (
...
+				if err != nil {
+					return fmt.Errorf("failed to marshal proto: %v", err)
+				}
+				update.Val = &gpb.TypedValue{Value: &gpb.TypedValue_AnyVal{AnyVal: any}}
+			} else {
+				b, err := json.Marshal(val)
+				if err != nil {
+					return fmt.Errorf("failed to marshal json: %v", err)
+				}
+				update.Val = &gpb.TypedValue{Value: &gpb.TypedValue_JsonVal{JsonVal: b}}
```
Comment (wenovus) [inline ygnmi/gnmi.go:373]: I'm not quite sure about doing this -- if the input path's origin is "openconfig" or "" (which is implicitly openconfig), JsonVal should never be used. I'd suggest the entire fallback routine to only apply when the origin is not openconfig.

Comment (DanG100) [reply ygnmi/gnmi.go]: added check for origin

File: ygnmi/gnmi.go [hunk 4]
```diff
@@ -330,25 +333,49 @@ const (
...
+		if opt.preferProto {
+			typedVal, err = ygot.EncodeTypedValue(val, gpb.Encoding_JSON_IETF, &ygot.RFC7951JSONConfig{AppendModuleName: true, PreferShadowPath: preferShadowPath})
+		} else {
+			typedVal = &gpb.TypedValue{Value: &gpb.TypedValue_JsonIetfVal{}}
+			// Since the GoStructs are generated using preferOperationalState, we
+			// need to turn on preferShadowPath to prefer marshalling config paths.
+			typedVal.Value.(*gpb.TypedValue_JsonIetfVal).JsonIetfVal, err = ygot.Marshal7951(val, ygot.JSONIndent("  "), &ygot.RFC7951JSONConfig{AppendModuleName: true, PreferShadowPath: preferShadowPath})
+		}
+
+		if err != nil && opt.setFallback && path.Origin != "openconfig" {
```
Comment (wenovus) [inline ygnmi/gnmi.go:357]: Check origin is not ""

Comment (DanG100) [reply ygnmi/gnmi.go:357]: resolvePath always sets the origin to "openconfig" if it empty https://github.com/openconfig/ygnmi/blob/main/ygnmi/gnmi.go#L410

Comment (wenovus) [reply ygnmi/gnmi.go:357]: The current unit test seems to interpret "" as non-openconfig.

Comment (DanG100) [reply ygnmi/gnmi.go:357]: fixed and added test

**Review by wenovus — APPROVED** [review]

**Pull request merged**